src/jail_rules.c

```diff
@@ -0,0 +1,933 @@
+{
+    /*
+     * Allocate the whole thing from a single memory block, so that a NULL
+     * terminated array of these things can be deallocated with g_strfreev.
+     */
+    gsize len = strlen(name);
+    JailDBusName* dbus_name = g_malloc(G_ALIGN8(sizeof(*dbus_name) + len + 1));
+    char* dest = (char*)dbus_name + G_ALIGN8(sizeof(*dbus_name));
+
+    memcpy(dest, name, len + 1);
+    dbus_name->name = dest;
+    dbus_name->require = require;
+    return dbus_name;
+}
+
+static
+void
+jail_rules_parse_dbus_names(
+    GKeyFile* keyfile,
+    const char* section,
+    const char* key,
+    GPtrArray* list)
+{
+    char* val = g_key_file_get_string(keyfile, section, key, NULL);
+
+    if (val) {
+        char** vals = g_strsplit_set(val, ",", -1);
+        char** ptr = vals;
+
+        for (ptr = vals; *ptr; ptr++) {
+            char* val = g_strstrip(*ptr);
+
+            if (g_dbus_is_name(val)) {
+                /* Consider all of them as required for now */
+                g_ptr_array_add(list, jail_rules_dbus_name_new(val, TRUE));
+            }
+        }
+
+        g_strfreev(vals);
+        g_free(val);
+    }
+}
+
+static
+void
+jail_rules_restrict_dbus_names(
+    const JailDBusName* const* in,
+    GPtrArray* out,
+    const char* const* names)
+{
+    if (in) {
+        int i;
+
+        for (i = 0; in[i]; i++) {
+            const JailDBusName* dbus = in[i];
+
+            if (dbus->require ||
+                gutil_strv_contains((GStrV*)names, dbus->name)) {
+                g_ptr_array_add(out, jail_rules_dbus_name_new
+                    (dbus->name, dbus->require));
+            }
+        }
+    }
+}
+
+static
+JailRulesData*
+jail_rules_data_alloc(
+    void)
+{
+    JailRulesData* data = g_new0(JailRulesData, 1);
+
+    data->permits = g_ptr_array_new_with_free_func(g_free);
+    data->profiles = g_ptr_array_new_with_free_func(g_free);
+    data->paths = g_ptr_array_new_with_free_func(g_free);
+    data->dbus_user_own = g_ptr_array_new_with_free_func(g_free);
+    data->dbus_user_talk = g_ptr_array_new_with_free_func(g_free);
+    data->dbus_system_own = g_ptr_array_new_with_free_func(g_free);
+    data->dbus_system_talk = g_ptr_array_new_with_free_func(g_free);
+    return data;
+}
+
+static
+JailRulesData*
+jail_rules_data_new(
+    const JailConf* conf)
+{
+    JailRulesData* data = jail_rules_data_alloc();
+
+    /* Always include required base profile */
+    jail_rules_add_profile(data, conf, SAILJAIL_BASE_PERM, TRUE);
+    return data;
+}
+
+static
+JailRulesData*
+jail_rules_parse_section(
+    GKeyFile* kf,
+    const char* section,
+    const JailConf* conf,
+    const JailRulesOpt* opt)
+{
+    JailRulesData* data = jail_rules_data_new(conf);
+    const char* app = opt->sailfish_app;
+    char* val;
+
+    /* Permissions= */ 
+    val = g_key_file_get_string(kf, section, SAILJAIL_KEY_PERMS, NULL);
+    if (val) {
+        JAIL_PERMIT_TYPE permit;
+        char** vals = g_strsplit_set(val, ":,", -1);
```

Comment:
Is there some reason not to use g_key_file_get_string_list() / allow list separator used by g_key_file_xxx? (= ";")?

Reply:
`g_key_file_get_string_list()` supports only one separator but (for some reason) I wanted to support more than one. I can add `;` to that list too. Or... Well, let me think about it.